### PR TITLE
Fix comma after struct update syntax

### LINKED ISSUE #41834: Comma after struct expansion is a syntax error
On rustc 1.17:

```rust
struct Foo {
    one: i32,
    two: i32,
    three: i32,
}

impl Default for Foo {
    fn default() -> Self {
        Foo { one : 1, two: 2, three: 3 }
    }
}

fn main() {
    // this works
    let foo = Foo {
        one: 11,
        ..Foo::default()
    };
    // this is a syntax error
    let foo = Foo {
        one: 111,
        ..Foo::default(),
    };
}
```

(playground link: https://is.gd/2wvenA)

This feels very inconsistent, as Rust has no problems with extra commas anywhere else. It is also an extremely "natural" mistake to make, so even if it's rightly signaled & easily corrected, it is still quite annoying.

### PR DESCRIPTION
**Pull request by F001**
Fix #41834 

r? @petrochenkov

### CODE REVIEW

**Comment by arielb1** [comment]
Thanks for the PR @F001! @petrochenkov (or someone else) will review your PR shortly. 

However, your PR does not compile because you have trailing whitespace in your test (error copied from travis CI log):
```
[00:02:01] tidy check (x86_64-unknown-linux-gnu)
[00:02:01] tidy error: /checkout/src/test/run-pass/issue-41834.rs:31: trailing whitespace
[00:02:02] some tidy checks failed
```
You'll have to fix that before the PR can be merged.

**Comment by F001** [comment]
Oh~  Thanks very much. Fixed.

**Comment by estebank** [comment]
@F001 what's the [new output of the following](https://is.gd/zwCAzd)?

```rust
    let _foo = Foo {
        one: 11,
        ..Foo::default(),
        three: 1
    };
```

r=me modulo comment.

**Comment by F001** [comment]
r? @estebank 

I tested this case and it compiles. The value of `_foo` is `Foo { one: 11, two: 2, three: 1 }`

And the case below can also compile, the result is `Foo { one: 110, two: 100, three: 120}`
```
let _foo = Foo {
        one: 110,
        ..Foo::default(),
        three: 120,
        two: 100,
    };
```

**Comment by F001** [comment]
But I find that struct update can now exist multiple times.

```
let _foo = Foo {
    one: 100,
    ..Foo::default(),
    ..Foo::new(),
};
```
And the below one will overwrite the upper one. 
But since the duplication of fields is an error, I tend to reject this kind of code. I'll find a way to fix it.

**Comment by F001** [comment]
Now, the struct update syntax can be specified only once. And it can be appear in any position in the initialization expression.

**Comment by petrochenkov** [comment]
Please do not merge yet.
The syntax needs to be harmonized with other `..`-like constructions and this is a sufficiently large change.
(I'll write more details tomorrow.)

**Comment by estebank** [comment]
@F001 I'm ok with the current output. I'll wait for your comments, @petrochenkov.

**Comment by petrochenkov** [comment]
TLDR: this PR reduces consistency. It's also a new feature (or even 2 features) and not just a fix.

There are several constructions that currently have the meaning "rest of the list" (ROTL):
```rust
S { field, ..base } // Struct expression
S { field, .. } // Struct pattern
S(field, .., field) // Tuple struct pattern
(field, .., field) // Tuple pattern
[elem, elems.., elem] // Slice pattern
[elem, .., elem] // Slice pattern (sugar for [elem, _.., elem])
fn f(arg, ...); // Variadic foreign function
```

All these constructions behave consistently in two regards:
1) Trailing separator is not permitted after ROTL
```rust
S { field, ..base, } // FAIL
S { field, .., } // FAIL
S(field, ..,) // FAIL
(field, ..,) // FAIL
[elem, elems..,] // FAIL
[elem, ..,] // FAIL
fn f(arg, ...,); // FAIL
```

2) If ROTL applies to the whole list (i.e. its position is irrelevant), then it's permitted only in the "conventional" last position.
```rust
S { field, ..base, field} // FAIL
S { field, .., field} // FAIL
fn f(arg, ..., arg); // FAIL
```

So, *if* some change is done to one of these syntaxes, other ones need to be updated as well.

However, I don't think the change this PR does *need* to be done.
ROTL is not part of the list, it's "one more thing in addition to the list". In this case the trailing separator is not useful for macros. It's also not useful for diffs (you are not going to add another `..base` after the first one). So the only remaining benefit is pretty subjective aesthetics.

There's one more thing - if you want to support `StructExpr { field, ..base, field}`, you'll have to change AST and HIR to track the exact position of `..base` and update trans to keept left-to-right evaluation order, otherwise `..base` will continue to be always evaluated last.

---

As a result, I think there's not enough motivation for this language extension.
Pinging @rust-lang/lang for their thoughts.

**Comment by F001** [comment]
Great summary. So many things I can learn. Thank you.

**Comment by aturon** [comment]
Nominated for lang team discussion.

**Comment by nikomatsakis** [comment]
[Cross-post from other thread](https://github.com/rust-lang/rust/issues/41834#issuecomment-304328283):

So we discussed this some in the @rust-lang/lang meeting. I would say that opinions were mixed, but in general I think that we are inclined (for now) to leave this as is. 

The arguments basically were as follows. First, everyone agreed that we actively do **not** want to allow `..Foo` anywhere but at the *end* of a struct expression.

Given that, the argument in favor of allowing commas is essentially that people might expect to put it there because they always end lines with a comma (habit, essentially). 

The argument *against* is that a trailing comma's main purpose is to avoid messy diffs and simplify macro authors lives, and @petrochenkov made a [convincing case that neither applies here](https://github.com/rust-lang/rust/pull/41848#issuecomment-301239194). Moreover, a trailing comma suggests that you *can* move things around in the list, and that you *might* put something afterwards, neither of which are true here. So including the comma is somewhat misleading.

Therefore, I think we're inclined to leave the grammar as is, and prohibit trailing commas after a `..Foo` syntax in a struct.

**Comment by nikomatsakis** [comment]
I'm going to close this PR for now -- any further discussion can take place on #41834.

**Comment by nikomatsakis** [comment]
(By the way, thanks to @F001 for jumping on this in any case!)
